### PR TITLE
ENT-4016: Fix copy_from overwriting permissions on destination file

### PR DESCRIPTION
**Pull request by karlhto**
Depends on the `preserve` attribute, although we should probably discuss how it should look @nickanderson 

Need to run acceptance tests.

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a pull request! Maybe @olehermanse can review this?

**Comment by karlhto** [comment]
@cf-bottom Jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2751/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2751/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2751/

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2757/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2757/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2757/

File: tests/acceptance/10_files/02_maintain/004.cf [hunk 1]
```diff
@@ -49,6 +49,7 @@ body perms test_og
 
 body copy_from test_copy
 {
+      preserve => "true";
```
Comment (nickanderson) [inline tests/acceptance/10_files/02_maintain/004.cf]: OK, I reviewed this one. I think it's ok to change this test here to specify mode true. The policy was expecting the source and target modes to match, and it determined the expected mode by querying the source file for its mode.

File: tests/acceptance/10_files/02_maintain/005.cf [hunk 2]
```diff
@@ -47,6 +47,7 @@ body perms test_og
 
 body copy_from test_copy
 {
+      preserve => "true";
```
Comment (nickanderson) [inline tests/acceptance/10_files/02_maintain/005.cf]: - Similar to 004 testfiles expected mode is based on queried mode for mocked
    etc/group source file.
    - So the test expected preservation of the mode, but it was not specified.

File: tests/acceptance/10_files/02_maintain/006.cf [hunk 3]
```diff
@@ -49,6 +49,7 @@ body perms test_og
 
 body copy_from test_copy
 {
+      preserve => "true";
```
Comment (nickanderson) [inline tests/acceptance/10_files/02_maintain/006.cf]: - Similar to 004 and 005 testfiles expected mode is based on queried mode for
    mocked etc/group source file.
    - So the test expected preservation of the mode, but it was not specified.

**Comment by karlhto** [comment]
Removed an `fprintf` statement, the above tests still count.

Also, I think the commits are fine. :^)

File: cf-agent/verify_files_utils.c [hunk 4]
```diff
@@ -1577,8 +1578,19 @@ bool CopyRegularFile(EvalContext *ctx, const char *source, const char *dest, con
         }
     }
     else
-    {
 #endif
+    {
+        // Retain destination file's permissions
+        if (!attr->copy.preserve)
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: I'd add a comment here that `attr->copy.preserve` actually means that the permissions of the **origin** should be preserved. Otherwise this is quite confusing.

File: cf-agent/verify_files_utils.c [hunk 5]
```diff
@@ -1577,8 +1578,19 @@ bool CopyRegularFile(EvalContext *ctx, const char *source, const char *dest, con
...
+        {
+            int mode = backupok ? (s.st_mode & 0777) : CF_PERMS_DEFAULT;
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: What does the mode have to do with `backupok`? And what is this `& 0777` supposed to limit? SUID bit?

File: cf-agent/verify_files_utils.c [hunk 6]
```diff
@@ -507,13 +507,8 @@ static PromiseResult CfCopyFile(EvalContext *ctx, char *sourcefile,
...
         {
-            if (((S_ISDIR(dsb.st_mode)) && (!S_ISDIR(srcmode))) ||
-                ((S_ISREG(dsb.st_mode)) && (!S_ISREG(srcmode))) ||
-                ((S_ISBLK(dsb.st_mode)) && (!S_ISBLK(srcmode))) ||
-                ((S_ISCHR(dsb.st_mode)) && (!S_ISCHR(srcmode))) ||
-                ((S_ISSOCK(dsb.st_mode)) && (!S_ISSOCK(srcmode))) ||
-                ((S_ISFIFO(dsb.st_mode)) && (!S_ISFIFO(srcmode))) ||
-                ((S_ISLNK(dsb.st_mode)) && (!S_ISLNK(srcmode))))
+            // Compare file type, check manpage inode(7) for explanation
+            if ((dsb.st_mode & S_IFMT) != (srcmode & S_IFMT))
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: Nice!

File: cf-agent/verify_files_utils.c [hunk 7]
```diff
@@ -591,11 +586,8 @@ static PromiseResult CfCopyFile(EvalContext *ctx, char *sourcefile,
                          sourcefile, destfile);
                     result = PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
                 }
-
-                return result;
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: Do you think this early `return` should really be removed? It might make things easier to read when there's a lot of code below like in this case.

File: cf-agent/verify_files_utils.c (hunk 5 above)
Comment (karlhto) [reply cf-agent/verify_files_utils.c]: Well, it's not well explained. `backupok` means that there was a file that will be overwritten (but there was a backup). If there was no file, the mode is set to `CF_PERMS_DEFAULT`.

I think I might work a bit more on this PR and create a new or modify the existing copy function that also takes permissions so I don't change the mode _after_ the file is created. Just to get rid of a bit of technical debt.

Comment (karlhto) [reply cf-agent/verify_files_utils.c]: > And what is this `& 0777` supposed to limit? SUID bit?

And yeah, was thinking the SUID bit.

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2879/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2879/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2879/

File: libpromises/files_copy.c [hunk 8]
```diff
@@ -36,6 +36,67 @@
...
+    {
+        Log(LOG_LEVEL_INFO,
+            "Unable to open destination file while copying '%s' to '%s'"
+            " (open: %s)", source, destination, GetErrorStr());
+        close(sd);
+        return false;
+    }
+
+    size_t total_bytes_written;
+    bool   last_write_was_hole;
```
Comment (vpodzime) [inline libpromises/files_copy.c]: I'd drop the extra spaces here.

File: libpromises/files_copy.c [hunk 9]
```diff
@@ -36,6 +36,67 @@
...
+    bool ret = FileSparseCopy(sd, source, dd, destination,
+                              ST_BLKSIZE(statbuf),
+                              &total_bytes_written, &last_write_was_hole);
+    if (!ret)
+    {
+        unlink(destination);
+        close(sd);
+        return false;
```
Comment (vpodzime) [inline libpromises/files_copy.c]: Should `close(dd)` here too, right?

File: libpromises/files_copy.c [hunk 10]
```diff
@@ -36,6 +36,67 @@
...
+    }
+
```
Comment (vpodzime) [inline libpromises/files_copy.c]: `close(sd)` here.

File: libpromises/files_copy.c [hunk 11]
```diff
@@ -36,6 +36,67 @@
...
+    bool do_sync = false;
+    ret = FileSparseClose(dd, destination, do_sync,
+                          total_bytes_written, last_write_was_hole);
+    if (!ret)
+    {
+        unlink(destination);
+        close(sd);
```
Comment (vpodzime) [inline libpromises/files_copy.c]: `close(dd)`

File: libpromises/files_copy.c [hunk 12]
```diff
@@ -36,6 +36,67 @@
 #include <string_lib.h>
 #include <acl_tools.h>
 
+bool CopyRegularFileDiskPerms(const char *source, const char *destination,
```
Comment (vpodzime) [inline libpromises/files_copy.c]: *tha same* in the commit message

File: cf-agent/verify_files_utils.c [hunk 13]
```diff
@@ -1345,20 +1345,22 @@ bool CopyRegularFile(EvalContext *ctx, const char *source, const char *dest, con
...
             Log(LOG_LEVEL_ERR, "Unable to construct filename for copy");
             return false;
         }
+    }
 
-#ifdef __APPLE__
+    bool dest_exists = false;
+    if (stat(dest, &dest_stat) == 0)
+    {
+        dest_exists = true;
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: I'd do `ret = stat(dest, &dest_stat); dest_exists = (ret == 0)` here because `dest_stat` is used later and is not just populated as a side effect of checking if the file exists or not.

File: cf-agent/verify_files_utils.c [hunk 14]
```diff
@@ -1374,100 +1376,120 @@ bool CopyRegularFile(EvalContext *ctx, const char *source, const char *dest, con
...
+                cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, attr,
+                     "Failed copying file '%s' to '%s'", source, new);
+                *result = PromiseResultUpdate(*result, PROMISE_RESULT_FAIL);
+                return false;
+            }
+        }
+        else
+        {
+            int mode =
+                dest_exists ? (dest_stat.st_mode & 0777) : CF_PERMS_DEFAULT;
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: This can just be a on a single line I think.

Comment (vpodzime) [reply cf-agent/verify_files_utils.c]: Maybe add a comment like `/* never preserve SUID bit */`?

File: cf-agent/verify_files_utils.c [hunk 15]
```diff
@@ -1281,8 +1281,8 @@ bool CopyRegularFile(EvalContext *ctx, const char *source, const char *dest, con
     char backup[CF_BUFSIZE];
     char new[CF_BUFSIZE], *linkable;
     int remote = false, backupisdir = false, backupok = false, discardbackup;
-    struct stat s;
-    struct stat dstat;
+    struct stat dest_stat;
+    struct stat new_stat;
```
Comment (vpodzime) [inline cf-agent/verify_files_utils.c]: These declarations should be moved close to the place these variables are used for the first time.

File: libpromises/files_copy.c [hunk 16]
```diff
@@ -124,7 +124,7 @@ bool CopyRegularFileDisk(const char *source, const char *destination)
     unlink(destination);
 
     int dd = safe_open_create_perms(destination,
-                                    O_WRONLY | O_CREAT | O_TRUNC | O_EXCL | O_BINARY,
+                                    O_WRONLY | O_CREAT | O_EXCL | O_BINARY,
```
Comment (vpodzime) [inline libpromises/files_copy.c]: This looks like an arbitrary/unrelated change.

File: libpromises/files_copy.c (hunk 12 above)
Comment (karlhto) [reply libpromises/files_copy.c]: Whoops.

File: cf-agent/verify_files_utils.c (hunk 15 above)
Comment (karlhto) [reply cf-agent/verify_files_utils.c]: Agree. :+1:

File: libpromises/files_copy.c (hunk 16 above)
Comment (karlhto) [reply libpromises/files_copy.c]: Well, it's just cleanup. `O_TRUNC` does nothing together with `O_CREAT` and `O_EXCL`.

**Comment by olehermanse** [comment]
@cf-bottom let's try in jenkins now!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2894/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2894/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2894/

File: libpromises/files_copy.c [hunk 17]
```diff
@@ -141,8 +141,8 @@ bool CopyRegularFileDisk(const char *source, const char *destination)
                          ST_BLKSIZE(statbuf),
                          &total_bytes_written, &last_write_was_hole);
     bool do_sync = false;
-    ok2= FileSparseClose(dd, destination, do_sync,
-                         total_bytes_written, last_write_was_hole);
+    ok2 = FileSparseClose(dd, destination, do_sync,
+                          total_bytes_written, last_write_was_hole);
```
Comment (olehermanse) [inline libpromises/files_copy.c]: The problem (one of them) with binpack illustrated :)

File: libpromises/files_copy.c [hunk 18]
```diff
@@ -36,6 +36,68 @@
...
+{
+    int sd = safe_open(source, O_RDONLY | O_BINARY);
+    if (sd == -1)
+    {
+        Log(LOG_LEVEL_INFO, "Can't copy '%s' (open: %s)",
+            source, GetErrorStr());
+        return false;
+    }
+
+    /* We need to stat the file to get the right source permissions. */
```
[the remark this answers is not in the document]
Comment (karlhto) [reply libpromises/files_copy.c]: Whoops, I rushed a bit with this function.

I think adding `assert`s should also be done.

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2915/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2915/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2915/

**Comment by karlhto** [comment]
No major changes. Added `assert`s and changed the comment you mentioned @olehermanse

**Comment by karlhto** [comment]
Cherry-pick: https://github.com/cfengine/core/pull/4006
